### ISSUE TITLE
Add support for parsing argument of an argument for attributes

### ISSUE THREAD

**Issue by kayagokalp** [FuelLabs/sway#3576]
To unblock https://github.com/FuelLabs/sway/issues/3544 we need to be able to somehow specify a revert code with `#[test(should_revert)]` annotations.  This requires some parser tweak and I have some possible options in my mind but it would be nice to discuss and get it in separately.

I think `should_revert` part of the `#[test(should_revert)]`  is called argument so basically I need a way to add an argument to an argument 😅 

### To specify a revert code of 0
```rust
#[test(should_revert(0))]
fn test_foo() {

}
```
### To specify any revert code (parser allows this now)
```rust
#[test(should_revert)]
fn test_foo() {

}
```

### To specify a revert code of 0
```rust
#[test(should_revert = "0")]
fn test_foo() {

}
```

### To specify any revert code (parser allows this now)
```rust
#[test(should_revert)]
fn test_foo() {

}
```

I think both options requires a little tweaking with the parser which shouldn't be a problem but curios about your thoughts cc @mitchmindtree @mohammadfawaz. 
I am leaning towards option 2 as Rust also has it, example:
```rust
#[cfg(target_os = "linux")]
```

**Comment by mohammadfawaz** [comment]
What about going full Rust and having 
```
    #[should_revert(expected = "Revert(0)")]
```
which similar to 
![image](https://user-images.githubusercontent.com/59666792/207128958-09c7b3d4-214d-42ee-a4f0-c9db170b28d6.png)

**Comment by mohammadfawaz** [comment]
> I will try that but I am assuming rust does that check.

I would imagine!

**Comment by mitchmindtree** [comment]
I think I still prefer the original proposal of allowing for recursively "nesting" of attributes.

> What about going full Rust

Rust's support for both nesting *and* assignment always struck me as a little unnecessary, and it's not always clear which to choose when designing macro APIs in Rust. E.g.

```rust
#[test(should_panic(expected("Revert(0)")))]
```
```rust
#[cfg(target_os("linux"))]
```

I think these are just as readable and clear, and both require less syntax and keep the macro model a little simpler. 

> In that case we might want to add a check to make sure that#[should_panic] is not used without #[test].

I think this is a good argument against using a totally separate attribute for `should_revert`. If we require that the attribute is nested within a test attribute (as we currently do), it's easier to check it's used within a valid context, and it's easier for us to work with the whole set of test-related attributes within a single structure (rather than having to check separately for `test` and `should_revert`).

> Add support for parsing argument of an argument

Rather than manually parsing an "argument of an argument", my intuition would be to represent attributes with a recursive data structure like a [rose tree](https://en.wikipedia.org/wiki/Rose_tree).

**Comment by kayagokalp** [comment]
Is this resolved by #4134 cc @AlicanC? I guess option 2 in the original issue is possible now?

**Comment by sdankel** [comment]
Closing as `#[test(should_revert = "42")]` is supported now.
